Attempt to improve gui design

blacksponge opened the pull request:
## Purpose

Try to improve the overall look and feel and accessibility of the GUI.

## Context



## Changes

Highlights :

* Make the "about" lines shorter by making this area double columns and adding an decorative picture
* Improve color contrasts and make color usage more uniform
* Group inputs in fieldset and add visible label for all fields in advanced form
* Change FAQ layout: This change needed to move the content of the FAQ into the code and integrate it to the "Angular translation system" instead of having separate markdown files. This means that translated FAQ will need to be imported, knowing that I have changed some wording at the same time. ~~Also I took this oportunity to test crowdin.com integration, <https://crowdin.com/project/zonemaster-test>.~~
* Slightly bigger font size and line height
* Make document outline more consistent.
* Replace links that looks like buttons in result filter area
* Add labels and group fields in result filter area, add expand + collapse buttons

<details>
<summary>My todo lists</summary>

Things I want to do (in this PR)
- [x] Make "about" lines shorter
- [x] Improve contrast between active and not active navigation links
- [x] Make color usage more uniform
  - [x] disabled primary button color
  - [ ] secondary button color -- change or keep ?
  - [x] lighten danger and secondary button on hover instead of darkening them
- [x] Try to improve "about" information visualization in "about" text (use markup)
  - [x] About zonemaster
  - [x] ~~About DNS~~ need to change the text, not doing this now
- [x] Improve accessibility of advanced option form:
  - [x] make labels visible
  - [x] use a fieldset to group NS and DS inputs
- [x] Improve accessibility of result filtering form
  - [x] maybe replace links that look like buttons with checkboxes and labels in fieldset
  - [x] Move help / severity level info out of tooltips
- [x] Improve document outline
  - [x] replace h4/h5 with h2/h3 in about section
  - [x] Add section + headings for form  / result in result page
- [x] Improve "program version" (at least make the "program version" thing a link so it is an actionable item)
- [x] ~~Expand (+ highlight?) text in search result~~ not in this PR
- [x] Change FAQ layout
  - [ ] add "expand" button for each section -- surprisingly hard to do, need more thoughts

Things that I broke and need to fix
- [x] fix advanced form NS and DS inputs wrapping for medium / small screens
- [x] main heading for result page
- [x] fix message banner
- [x] Click on faq entry update url fragment
- [x] change of url fragment open faq entry
- [x] fix link in info alert in advanced form
- [x] i18n (french translation done)
- [x] fix advanced switch (transition too long for focus outline, colors don't match other secondary buttons)
</details>

## How to test this PR

A live version is deployed at https://zonemaster.fr

// TODO

matsduf commented:
@blacksponge, have you considered #315 ("Put lists of IP addresses in messages under a clickable label") in this PR?

blacksponge commented:
I have, but it is actually out of scope. I am not touching the results details in this one.
I do plan to improve it in the future, to also address the point `Expand (+ highlight?) text in search result` that I have crossed out of the todo list for now.

matsduf commented:
I will review when this is no longer a draft.

blacksponge commented:
> I will review when this is no longer a draft.

This is now ready for review.

matsduf commented:
Do these changes make it easier to customize the look and feel? My colleague has asked for the CSS to be decoupled from the code to make it easier to change CSS. Will these changes make it easier or harder to change CSS?

For me it looks like the FAQ is integrated in HTML code. How do you expect translation to be done?

matsduf commented:
I changed language in the FAQ to Swedish and after that whatever I do I end up in "file not found". Not even reloading helps.

blacksponge commented:
> I changed language in the FAQ to Swedish in zonemaster.fr and after that whatever I do I end up in "file not found". Not even reloading helps.

Only English and French are deployed on zonemaster.fr.

blacksponge commented:
> For me it looks like the FAQ is integrated in HTML code. How do you expect translation to be done?

Using the same translation system as all other strings in the GUI.

blacksponge commented:
> Do these changes make it easier to customize the look and feel? My colleague has asked for the CSS to be decoupled from the code to make it easier to change CSS. Will these changes make it easier or harder to change CSS?

No. It is still Angular, so no change on that side.

matsduf commented:
> > For me it looks like the FAQ is integrated in HTML code. How do you expect translation to be done?
> 
> Using the same translation system as all other strings in the GUI.

That works for strings, but maybe less good for texts. A reply is often more than one paragraph. I will do a test translation as part of reviewing. Do you see that it will be possible to migrate current translations into the new framework?

blacksponge commented:
> That works for strings, but maybe less good for texts. 

I would agree but having it in markdown is very non optimal. It limits the document structure. For the kind of things that I have done it would either mean writing directly html or a custom markdown parser. Both would be hard to integrate within the current framework. And I am not please with either of the current solution, whether it is requesting a static html file to inject in the page or having a full angular app while all we really need is a static html file and 15 lines of javascript.

> I will do a test translation as part of reviewing.

As said in the description I am also testing Crowdin for translation. If you want I can add you to my test project and you can test that as well. 

> Do you see that it will be possible to migrate current translations into the new framework?

Partial manual migration is possible, there have been some small changes in a few strings but only a minority.

blacksponge commented:
> > For me it looks like the FAQ is integrated in HTML code. How do you expect translation to be done?
> 
> Using the same translation system as all other strings in the GUI.

My mistake, it looks like I have overridden the configuration in the last deployment, it should not have displayed Swedish.

matsduf commented:
It is not possible for me to view zonemaster.fr on some platforms. It seems like if the preferred language is "sv" it will end up trying to load a page not existing. Can that be changed?

matsduf commented:
As far as I can see, the pictures on the first page are hard-coded into the GUI. I think that is wrong. If there should be room for picture, which could be fine, it should be easy to exchange those for other pictures that fits the frame that the GUI belongs to. The pictures, if any, should be se in app.config.json.

The PR also changes the texts, as far as I can see. It should be separated so that we decide on form and text in different steps. That will also make it easier to see what texts have changed for the translation work.

blacksponge commented:
> The PR also changes the texts, as far as I can see. It should be separated so that we decide on form and text in different steps. That will also make it easier to see what texts have changed for the translation work.

I changed some of the text indeed, I can move the FAQ changes outside of this PR but I will keep the "About Zonemaster" in this one.

blacksponge commented:
> > As far as I can see, the pictures on the first page are hard-coded into the GUI. I think that is wrong. If there should be room for picture, which could be fine, it should be easy to exchange those for other pictures that fits the frame that the GUI belongs to. The pictures, if any, should be se in app.config.json.
> 
> I can do that.

As the pictures are definied in CSS and that the style can be overridden with #439 I do not see the need to add a specific configuration.

blacksponge commented:
> > The PR also changes the texts, as far as I can see. It should be separated so that we decide on form and text in different steps. That will also make it easier to see what texts have changed for the translation work.
> 
> I changed some of the text indeed, I can move the FAQ changes outside of this PR but I will keep the "About Zonemaster" in this one.

The FAQ changes in this PR have been reverted and moved to #443.

pnax commented on src/app/components/faq/faq.component.html:
```diff
@@ -1,6 +1,197 @@
+      Secondly, all tests that Zonemaster runs are defined in Test Case specifications that can be found in the <a href="https://github.com/zonemaster/zonemaster/tree/master/docs/public/specifications/tests#list-of-defined-test-cases">Defined Test Cases</a> document.
+    </p>
+    <p i18n="faq answer">
+      Thirdly, Zonemaster can be used to test undelegated domain names. See <a href="#what-is-an-undelegated-domain-test">Question 12</a>.
```
There is no more question number. Suggestion to use the question instead.
`See "What is an undelegated domain test?"`

blacksponge replied:
The wording has been changed in #443

pnax commented on src/app/components/faq/faq.component.html:
```diff
@@ -1,6 +1,197 @@
+      Thirdly, Zonemaster can be used to test undelegated domain names. See <a href="#what-is-an-undelegated-domain-test">Question 12</a>.
+    </p>
+    <p i18n="faq answer">
+      Fourthly, Zonemaster can be used to test DS records before their publication in the parent zone (which is required to enable DNSSEC for a signed zone). See <a href="#can-i-test-the-ds-records-before-they-are-published">Question 13</a>.
```
Same for `Question 13`. Suggestion:
`See "Can I test the DS records before they are published?"`.

Same comment on the link.

blacksponge replied:
The wording has been changed in #443

pnax commented on src/app/components/faq/faq.component.html:
```diff
@@ -1,6 +1,197 @@
+      Secondly, all tests that Zonemaster runs are defined in Test Case specifications that can be found in the <a href="https://github.com/zonemaster/zonemaster/tree/master/docs/public/specifications/tests#list-of-defined-test-cases">Defined Test Cases</a> document.
+    </p>
+    <p i18n="faq answer">
+      Thirdly, Zonemaster can be used to test undelegated domain names. See <a href="#what-is-an-undelegated-domain-test">Question 12</a>.
```
Weird the link is not behaving correctly. I don't know why. I can reproduce on https://zonemaster.fr/fr/faq#what-makes-zonemaster-different . The link leaves the FAQ page which is not the expected behavior.

blacksponge replied:
Ah yes, that is because of the base href, it should be fixed now, thanks for noticing.

pnax reviewed (commented):
Great work! This is a very nice GUI that you suggest. I've read both the landing page and the FAQ (in English only) and I've found some small typos/errors. Besides that I really like the way you improve the FAQ wording!

blacksponge commented on src/app/components/faq-question/faq-question.component.css:
```diff
@@ -0,0 +1,17 @@
+}
+
+.faq-question h3 {
+  display: inline-block;
```
Good catch, I have applied you suggestion.

blacksponge commented on src/app/components/faq/faq.component.html:
```diff
@@ -1,6 +1,197 @@
+      What kind of queries does Zonemaster generate?
+    </ng-container>
+    <p i18n="faq answer">
+      Zonemaster send multiple DNS queries to the name servers hosting the domain name being tested and also to the name servers hosting the parent zone of that domain name.
```
Fixed in #443

blacksponge commented on src/app/components/faq/faq.component.html:
```diff
@@ -1,6 +1,197 @@
+    <p i18n="faq answer">The Zonemaster tool could help different kind of users:</p>
+
+    <ul>
+      <li i18n="faq answer">Users who are knowledgable about the DNS protocol.</li>
```
The wording has been changed in #443

pnax commented on src/locale/messages.fr.xlf:
```diff
@@ -401,14 +335,474 @@
+      <trans-unit id="87d46545583105db21fb77d7385e902d661c2df0" datatype="html">
+        <source>What is the meaning of the severity levels?</source>
+        <note priority="1" from="description">result info</note>
+        <target state="translated">Que signifie les niveaux de sévérité ?</target>
```
I would use a plural form: `signifie` -> `signifient`

pnax reviewed (approved):
Beside a suggestion in the French translation I think this provides nice and welcome improvements to the GUI.

blacksponge commented:
Things that I have notice when zooming at 200% (text only on firefox)

* Overlap issue: The language icon size is set to 27px, need to be changed to 1em to be responsive

Some potential improvements, could be fixed if the behaviour followed the one defined for normal zoom (media query on screen size)

* In the result page, wrap the "Expand / collapse" buttons after the search
* In the result page, align the "History / Export / Share" buttons to the left when wrapping
* In the advanced form, wrap the inputs the same as normal zoom / mobile

blacksponge commented:
Other things to look at:
* overflow of the history modal (zoomed text)
* bad order of element in main form ("Run test" is before "Option", should be after)
* add title to inputs for nameservers and DS
* add aria-invalid for invalid inputs, required/aria-required for required inputs (only one?)

blacksponge commented:
UI issue : "delete" button is not correctly aligned when an error is displayed in NS/DS rows.

blacksponge commented:
So as I have noticed some issues with the current UI and this PR, I will fix the small issues here and properly open new issues for the more complex ones. Otherwise this PR will never be merged.

blacksponge commented on src/app/components/form/form.component.css:
```diff
@@ -26,17 +49,137 @@
+  .domain .form-section fieldset:first-child .delete-button-container {
+    margin-top: 2.2rem;
+    display: flex;
+    align-items: center;
```
Maybe better `align-items: baseline`

blacksponge commented:
Okay I am done adding more stuff to this PR. I will open issue for the remaining things.

matsduf commented:
Underlining signals link on the web, but the underlined text is not a link (there are more such cases):

<img width="361" alt="image" src="https://github.com/zonemaster/zonemaster-gui/assets/5430733/6a7ba42d-628f-492e-86ff-ea68c6935b69">

It is recommended not to use underlining when it is not a link.

matsduf commented:
If you select "History" you get a list of tests. The items look like links, but when you click on them nothing appearing happens (things are changed in the background, but that is hard to see).

matsduf commented:
What is the license and copyright of the pictures on the entry page? Is it stated somewhere?

They are hard coded in src/app/components/domain/domain.component.css. Is that the only place to change if you want to use something else?

matsduf commented:
We should rather refer to https://github.com/zonemaster/zonemaster/blob/master/docs/contact-and-mailing-lists.md then linking to the list server.

<img width="924" alt="image" src="https://github.com/zonemaster/zonemaster-gui/assets/5430733/8e28d5c6-4db2-48a2-a460-d3456333504f">

blacksponge commented:
* The line under the h2 has been removed
* A mention to the image license has been added to the readme
* The default contact address has been changed to zonemaster@

blacksponge commented:
test have been updated, I just need to update the screenshots for this to be fully ready

blacksponge commented:
> The problem is that it will lead to a lot of spam. The contact@ should work in the sense that you get a pointer to the page about mailing lists. I have asked IT why it does not. The risk is that we will have to do the same thing with the zonemaster@ address.

Is it working though ? Because from yesterday's tests it did not seem to work. And putting a non working email address is kind of misleading.

matsduf commented:
> > The problem is that it will lead to a lot of spam. The contact@ should work in the sense that you get a pointer to the page about mailing lists. I have asked IT why it does not. The risk is that we will have to do the same thing with the zonemaster@ address.
> 
> Is it working though ? Because from yesterday's tests it did not seem to work. And putting a non working email address is kind of misleading.

It is actually working when I just tested. In the working group we agreed to have it working like this, and using the adresse as the outward facing email address. 

I got the following:
```
Date: Fri, 8 Dec 2023 09:03:41 +0000
From: contact zonemaster <contact@zonemaster.net>
To: Mats Dufberg 
Subject: Automatic reply: Testmeddelande contact

This mail address has been closed to excessive amount of spam messages.
Please follow the link below for how to contact the Zonemaster project.

Cette adresse e-mail a été fermée suite à un nombre excessif de messages
indésirables. Veuillez suivre le lien ci-dessous pour savoir comment
contacter le projet Zonemaster.

https://github.com/zonemaster/zonemaster/blob/master/docs/contact-and-mailing-lists.md


Thank you/Merci

Zonemaster Project
https://github.com/zonemaster/zonemaster/
https://zonemaster.net/
```

matsduf commented:
> * The line under the h2 has been removed

Yes, I could see that. That is better.

blacksponge commented:
> could you test sending an email to contact@?

I just sent a email to <contact@zonemaster.net> (from both my personal and work email) and did not receive any response yet.

blacksponge commented:
The tests are now passing. What should I do regarding the email address?

matsduf commented:
> The tests are now passing. What should I do regarding the email address?

I will make sure it works for you anyone as for me. Please send the mail addresses you sent from to my email.

blacksponge commented on README.md:
```diff
@@ -110,6 +110,9 @@ License
 This is free software under a 2-clause BSD license. The full text of the license can
 be found in the [LICENSE](LICENSE) file included in this respository.
 
+Images `src/assets/images/person_looking_at_computer.svg` and `src/assets/images/world_connected.svg`
+are taken from <https://undraw.co>, [full license](https://undraw.co/license).
+
```
Honestly I don't know. There was already no need to put a mention of the license in the first place.

matsduf replied:
Please note that I have approved.

blacksponge replied:
Okay thanks, I will merge this so I can move on with the other PRs, and if we see the need we can rewrite this section / move it to the license file.
